From 06a5a2225310cac0760fb2c4e42d84da675fad9a Mon Sep 17 00:00:00 2001
From: Marc-Aurel Zent <mzent@codeweavers.com>
Date: Thu, 30 Oct 2025 22:02:27 +0100
Subject: [PATCH 41/44] ntdll: Fix an off-by-one error in the pseudo-handle
 check for inproc syncs.

Also add an is_pseudo_handle() helper.
---
 dlls/ntdll/unix/sync.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/dlls/ntdll/unix/sync.c b/dlls/ntdll/unix/sync.c
index dce97dbaa49..eb2ac0413e5 100644
--- a/dlls/ntdll/unix/sync.c
+++ b/dlls/ntdll/unix/sync.c
@@ -442,6 +442,10 @@ static inline unsigned int inproc_sync_handle_to_index( HANDLE handle, unsigned
     return idx % INPROC_SYNC_CACHE_BLOCK_SIZE;
 }
 
+static BOOL is_pseudo_handle( HANDLE handle )
+{
+    return ((ULONG)(ULONG_PTR)handle >= 0xfffffffa);
+}
 
 static struct inproc_sync_cache_entry *cache_inproc_sync_obj( HANDLE handle, obj_handle_t inproc_sync, int fd,
                                                               enum inproc_sync_type type, unsigned int access )
@@ -451,6 +455,9 @@ static struct inproc_sync_cache_entry *cache_inproc_sync_obj( HANDLE handle, obj
     sigset_t sigset;
     int refcount;
 
+    /* don't cache pseudo-handles; waiting on them is pointless anyway */
+    if (is_pseudo_handle( handle )) return NULL;
+
     if (entry >= INPROC_SYNC_CACHE_ENTRIES)
     {
         FIXME( "too many allocated handles, not caching %p\n", handle );
-- 
2.51.2

